### PR TITLE
Qprofiler jun2019

### PR DESCRIPTION
**Pull request by ChristinaXu2017**
1. add attribute readCount to all <sequenceMetrics> under <bamMetrics><bam field>. Added unit tests, make sure these counts are consistence to <bamSummary>. 
2. add attribute file_checksum to <bamReport>, so NSG check can know whether two input BAM with same name are identical or not.

### CODE REVIEW

File: qprofiler2/src/org/qcmg/qprofiler2/util/CycleSummaryUtils.java [hunk 1]
```diff
@@ -9,8 +9,8 @@
 import htsjdk.samtools.Cigar;
 
 public class CycleSummaryUtils {
-	public static final Pattern BAD_MD_PATTERN = Pattern.compile("([ACGTN])");	
-	private static final Pattern badReadPattern = Pattern.compile("([.N])");	
+//	public static final Pattern BAD_MD_PATTERN = Pattern.compile("([ACGTN])");
```
Comment (holmeso) [inline qprofiler2/src/org/qcmg/qprofiler2/util/CycleSummaryUtils.java]: delete

File: qprofiler2/src/org/qcmg/qprofiler2/summarise/PositionSummary.java [hunk 2]
```diff
@@ -152,7 +156,7 @@ public void addPosition(final int position, String  rgid  ) {
 		
 		rgCoverages[ readGroupIds.indexOf(rgid)  ].increment(position / BUCKET_SIZE);
```
Comment (holmeso) [inline qprofiler2/src/org/qcmg/qprofiler2/summarise/PositionSummary.java]: what is the performance impact of looping through a list (`readGroupIds.indexOf(rgid)`) for each record?
If you don't want to use a map, would it be better to sort the array and call `binarySearch` instead?

File: qprofiler2/src/org/qcmg/qprofiler2/summarise/PositionSummary.java [hunk 3]
```diff
@@ -27,7 +27,7 @@
 	private final AtomicInteger min;
 	private final AtomicInteger max;
 	private final QCMGAtomicLongArray[] rgCoverages; // the coverage for each readgroup on that position   
-	private final QCMGAtomicLongArray coverage = new QCMGAtomicLongArray(512);  // total coverage on that position
+//	private final QCMGAtomicLongArray coverage = new QCMGAtomicLongArray(512);  // total coverage on that position
```
Comment (holmeso) [inline qprofiler2/src/org/qcmg/qprofiler2/summarise/PositionSummary.java]: delete

File: qprofiler2/src/org/qcmg/qprofiler2/util/CycleSummaryUtils.java (hunk 1 above)
Comment (ChristinaXu2017) [reply qprofiler2/src/org/qcmg/qprofiler2/util/CycleSummaryUtils.java]: done

File: qprofiler2/src/org/qcmg/qprofiler2/summarise/PositionSummary.java (hunk 2 above)
Comment (ChristinaXu2017) [reply qprofiler2/src/org/qcmg/qprofiler2/summarise/PositionSummary.java]: There were only a couple readGroups in one BAM, but now sometimes we received BAM with 30 read groups. O(log(N)) is much faster than O(N), especially we only sort the list once. 
OK, the code is updated

File: qprofiler2/src/org/qcmg/qprofiler2/summarise/PositionSummary.java (hunk 3 above)
Comment (ChristinaXu2017) [reply qprofiler2/src/org/qcmg/qprofiler2/summarise/PositionSummary.java]: done!

**Pull request merged**